Handle TermsEnum Strictly with DLS enabled

jkakavas opened the pull request:
This is work in progress regarding changing the behavior of the DocuementSubsetReader to forbid certain APIs when configured to do so.

jkakavas committed:
WIP

elasticmachine commented:
Pinging @elastic/es-security

jpountz commented on server/src/main/java/org/elasticsearch/index/fielddata/plain/SortedSetDVOrdinalsIndexFieldData.java:
```diff
@@ -127,7 +128,11 @@ public IndexOrdinalsFieldData loadGlobal(DirectoryReader indexReader) {
-        return GlobalOrdinalsBuilder.build(indexReader, this, indexSettings, breakerService, logger, scriptFunction);
+        if (indexReader instanceof FilterDirectoryReader){
+           indexReader = FilterDirectoryReader.unwrap(indexReader);
+        }
```
Can you add a comment to explain the reasoning? I think we should try to make it more robust by only unwrapping DLS readers. Maybe one way to do that would be to have a sub class of FilterDirectoryReader called something like `UnwrapForGlobalOrdsFilterDirectoryReader` (terrible name, suggestions welcome), make the DLS reader extend it, and do the `instanceof` call on this class here and only unwrap once as opposed to recursively? But that would require that we always wrap with the DLS reader before the FLS reader so that the FLS reader could not hide the DLS reader, I don't know whether that could be problematic.

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetReader.java:
```diff
@@ -217,4 +230,267 @@ Bits getWrappedLiveDocs() {
+        @Override
+        public void seekExact(long ord) throws IOException {
+            throw new UnsupportedOperationException();
+        }
```
It'b be nice to have a good error message in these UnsupportedOperationExceptions explaining why they are disallowed.

jpountz commented on x-pack/plugin/src/test/resources/rest-api-spec/test/security/authz/50_build_global_ordinals.yml:
```diff
@@ -0,0 +1,94 @@
+          - '{"index": {"_index": "test_index", "_id": "u7"}}'
+          - '{"user": "u7", "followers": ["u1"]}'
+          - '{"index": {"_index": "test_index", "_id": "u8"}}'
+          - '{"user": "u8", "followers": ["u3"]}'
```
can you leave a comment that we do two refreshed bulk requests in order to have multiple segments so that someone doesn't collapse them by mistake?

jpountz commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java:
```diff
@@ -77,6 +77,10 @@ private XPackSettings() {
 
+    /** Setting for enabling or disabling safe terms enum behavior. Defaults to false. */
+    public static final Setting<Boolean> STRICT_TERMS_ENUM = Setting.boolSetting("xpack.security.strict_terms_enum.enabled", true,
+        Setting.Property.NodeScope);
```
I wonder whether this should be a node or index setting.

jpountz reviewed (commented):
I like it in general, I think it's close. Let's also add REST tests for fuzzy queries to make sure they break, and wildcard/prefix queries which don't leak data but will break because of this change, we should make sure to return an informative error message in such cases?

jkakavas committed:
address feedback and add tests

jkakavas committed:
Merge remote-tracking branch 'origin/master' into dls-terms-fix

jkakavas committed:
Merge remote-tracking branch 'origin/master' into dls-terms-fix

jkakavas committed:
fix compilation and checkstyle

jkakavas committed:
Fix tests

jkakavas committed:
address feedback

jkakavas committed:
Wrap with the FLS FilteredReader first

So that we know that the external wrapper is a DLS one and unwrap
only this when needed

jkakavas committed:
add strict terms enum as non replicated setting

jkakavas committed:
use the setting, not the string (duh)
